src/Control/SessionHandler/FileSessionHandler.php

```diff
@@ -0,0 +1,346 @@
+        return Path::join(...[$this->baseDir, ...$subParts, FileSessionHandler::SESSION_FILE_PREFIX . $id]);
+    }
+
+    /**
+     * Take a numeric string from save_path configuration and make sure it's a PHP octal integer.
+     */
+    private function ensureValueIsOctal(string $value): int
+    {
+        if (strlen($value) < 4) {
+            $value = '0' . $value;
+        }
+        return octdec($value);
+    }
+
+    /**
+     * Check if a session file needs its permissions to be updated to match configured mode.
+     */
+    private function needsPermissionUpdate(#[SensitiveParameter] string $path): bool
+    {
+        // Silence PHP's warning, any problems here will already be logged by the calling method.
+        $rawPerms = @fileperms($path);
+        // If we had trouble reading the permissions, just assume it needs updating.
+        // Either we'll be able to set it right, or the calling method will fail and log.
+        if ($rawPerms === false) {
+            return true;
+        }
+        // See https://www.php.net/manual/en/function.fileperms.php#refsect1-function.fileperms-examples
+        $permsInOctal = octdec(substr(sprintf('%o', $rawPerms), -4));
+        return $permsInOctal !== $this->mode;
+    }
+
+    /**
+     * Get the exception message without leaking session IDs
+     */
+    private function getSafeExceptionMessage(Exception $exception, #[SensitiveParameter] ?string $id = null): string
+    {
+        if ($id) {
+            return str_replace($id, '<session_id>', $exception->getMessage());
+        }
+        $prefix = FileSessionHandler::SESSION_FILE_PREFIX;
+        return preg_replace('/(' . $prefix . ')[,\w\-]+/', '$1<session_id>', $exception->getMessage());
+    }
+
+    /**
+     * Check if a session file is expired.
+     *
+     * This method is necessary becuse garbage collection may not have run yet.
+     *
+     * @throws RuntimeException if the modified time of the file can't be read
+     */
+    private function isSessionExpired(#[SensitiveParameter] string $path): bool
+    {
+        // Silence PHP's warning as it's just noise at this stage. Proper logging is handled by the caller of this method.
+        $mTime = @filemtime($path);
+        $maxLifeInSeconds = $this->getLifetime();
+        if ($mTime === false) {
+            throw new RuntimeException('Could not read modified time of session file');
+        }
+        return $mTime < DBDatetime::now()->getTimestamp() - $maxLifeInSeconds;
+    }
+
+    /**
+     * Returns the cookie lifetime if it's non-zero, otherwise returns the garbage collection lifetime.
+     */
+    private function getLifetime(): int
+    {
+        $cookieLifetime = (int) Session::config()->get('lifetime');
+        if ($cookieLifetime) {
+            return $cookieLifetime;
+        }
+        return (int) ini_get('session.gc_maxlifetime');
+    }
+
+    private function logError(string $message): void
```

Comment:
Add a generic docblock

Reply:
We don't need a docblock for private methods (since downstream devs can't use them and API docs don't show them) unless the method signature isn't self explanatory.

If there's something about this method signature that's unclear please let me know so I can fix it in the signature, rather than with docs.